File: binaryTree.java


class binaryTree{
	private int data;
	private binaryTree left;
	private binaryTree right;

	public binaryTree insert(binaryTree root, int data){
	
		if(root == null){
			binaryTree temp = new binaryTree();
                	temp.left = null;
                	temp.right = null;
                	temp.data = data;
			root = temp;
		}else{
			if(root.data > data)
				root.left = insert(root.left, data);
			if(root.data < data)	
				root.right = insert(root.right, data);
		}
		return root;
	}
	public void display(binaryTree root){
		binaryTree temp = new binaryTree();
		temp = root;
		if(root == null) return ;

		if(root.left != null) display(root.left);

		System.out.println("data->" + root.data);

		if(root.right != null) display(root.right);
	}
	public void delete (binaryTree root, binaryTree par, int data){
		if(root.data == data) {
			if(root.left == null && root.right == null) {
				if(par.left == root ){
					par.left = null;
				}else if(par.right == root){
					par.right = null;
				}

			}else if(root.left == null && root.right != null || root.left != null && root.right == null){

				if(root.left == null){
					if(par.left == root){
                                		par.left = root.right;
						root = null;
                                	}else if(par.right == root){
                                		par.right = root.right;
						root = null;
                       		 }
				}else if(root.right == null){
                                	if(par.left == root){
                                		par.left = root.left;
						root = null;
                                	}else if(par.right == root){
                                		par.right = root.left;
						root = null;
                               		}

				}	
			}else if (root.left != null && root.right != null){
                        	binaryTree temp = new binaryTree();
                       	 	binaryTree temp1 = new binaryTree();
                        	temp = root.right;
                        	temp1 = root;
                        	int value = 0;

                        	while(temp.left != null){
                                	temp1 = temp;
                                	temp = temp.left;
                       		 }
                        	if(temp.right != null) {
					if( temp1.right == temp  ){
                                		value = temp.data;
                                		temp1.right = temp.right;
						temp  = null;
                               		}else{
						value = temp.data;
                                                temp1.left = temp.right;
                                                temp  = null;
					     }		
                       		}else if(temp.right == null){
                                	value = temp.data;
                               		temp1.right =  null;
                        	}
                        	root.data = value;
               }
	
		
		}else{
			if(root.data > data)
                                 delete(root.left, root, data);
                        if(root.data < data)
                                 delete(root.right, root, data);
		}
	}
	public static void main(String args[]){
		binaryTree root = new binaryTree();
		binaryTree obj = new binaryTree();
		binaryTree par = new binaryTree();
		root = null;
		par  = null; 	
		root = obj.insert(root, 5);
		root = obj.insert(root, 6);
		root = obj.insert(root, 1);
		root = obj.insert(root, 4);
		root = obj.insert(root, 8);
                root = obj.insert(root, 9);
		root = obj.insert(root, 10);
                root = obj.insert(root, 14);
		
		obj.display(root);
		obj.delete(root,par,5);
		System.out.println("after deletion");
		obj.display(root);
		
	}
} 

File: binaryTree.c
#include<stdio.h>
#include<stdlib.h>

struct tree{
	struct tree *right;
	struct tree *left;
	int data;

}*root;

void  insert(struct tree **root , int data){

	if(*root == NULL){
		struct tree *temp = malloc(sizeof(struct tree));
		temp->right = NULL;
		temp->left  = NULL;
		temp->data  = data;
		*root = temp;
		return ;	
	}else{
		if((*root)->data > data){
			 insert(&(*root)->left,data);
		}else if ((*root) ->data < data){
			 insert(&(*root)->right, data);
		}	

	     }
}		

void delete( struct tree *root, struct tree *root1, int data){
	if(root->data == data ){
		if(root->left == NULL && root->right == NULL){
			if(root1->left == root){
				root1->left = NULL;
				free(root);
			}else if(root1->right == root){
				root1->right = NULL;
				free(root);
			}	
		}else if((root->left != NULL && root->right == NULL) || ( root->right != NULL && root->left == NULL)){
	
			if(root->left == NULL){
		        	if(root1->left == root){
                                root1->left = root->right;
                                free(root);
                        	}else if(root1->right == root){
                                root1->right = root->right;
                                free(root);
                        }else if(root->right == NULL){
                                if(root1->left == root){
                                root1->left = root->left;
                                free(root);
                                }else if(root1->right == root){
                                root1->right = root->left;
                                free(root);
                        	}

		
		        }
		     }		
	       }else if (root->left != NULL && root->right != NULL){
			struct tree *temp = malloc(sizeof(struct tree));
			struct tree *temp1 = malloc(sizeof(struct tree));	
			temp = root->right;
			temp1 = root;
			int value;

			while(temp->left != NULL){
				temp1 = temp;	
				temp = temp->left;
			}
			if(temp->right != NULL){
				value = temp->data; 
				temp1->left = temp->right;
				free(temp);	
			}else if(temp->right == NULL){
				value = temp->data;
				free(temp);
				temp1->right =  NULL;
			}	
			root->data = value;	
	       }
	}else{
		if( root->data > data){
			delete(root->left, root, data);
		}else if (root->data < data){
			delete(root->right, root, data);
		}	
	}
}
void  displayLVR(struct tree *root){
	if(root == NULL){
		return ;
	}else{
		displayLVR(root->left);
		printf("%d\t",root->data);
		displayLVR(root->right);
	}
}

int main(){
	root = NULL;
	insert(&root,5);
	insert(&root,6);
	insert(&root,1);
	insert(&root,4);
	insert(&root,9);
	insert(&root,10);
	insert(&root,14);
	insert(&root,8);
	displayLVR(root);
	delete(root,NULL,6);
	displayLVR(root);
	return 0;
}
